compiler/lib/global_deadcode.ml

```diff
@@ -0,0 +1,435 @@
+    + They are returned; or
+    + They are applied to a function. 
+ *)
+let zero prog sentinal live_table =
+  let compact_vars vars =
+    let i = ref (Array.length vars - 1) in
+    while !i >= 0 && Var.equal vars.(!i) sentinal do
+      i := !i - 1
+    done;
+    Array.sub vars ~pos:0 ~len:(!i + 1)
+  in
+  let is_live v =
+    match Var.Tbl.get live_table v with
+    | Dead -> false
+    | _ -> true
+  in
+  let zero_var x = if is_live x then x else sentinal in
+  let zero_instr instr =
+    match instr with
+    | Let (x, e) -> (
+        match e with
+        | Block (start, vars, is_array) -> (
+            match Var.Tbl.get live_table x with
+            | Live fields ->
+                let vars =
+                  Array.mapi
+                    ~f:(fun i v -> if IntSet.mem i fields then v else sentinal)
+                    vars
+                  |> compact_vars
+                in
+                let e = Block (start, vars, is_array) in
+                Let (x, e)
+            | _ -> instr)
+        | Apply ap ->
+            let args = List.map ~f:zero_var ap.args in
+            Let (x, Apply { ap with args })
+        | _ -> instr)
+    | _ -> instr
+  in
+  let zero_block block =
+    (* Analyze block instructions *)
+    let body = List.map ~f:(fun (instr, loc) -> zero_instr instr, loc) block.body in
+    (* Analyze branch *)
+    let branch =
+      let last, loc = block.branch in
+      let last =
+        match last with
+        | Return x -> Return (zero_var x)
+        | _ -> last
+      in
+      last, loc
+    in
+    { block with body; branch }
+  in
+  let blocks = prog.blocks |> Addr.Map.map zero_block in
+  { prog with blocks }
+
+module Print = struct
+  let live_to_string = function
+    | Live fields ->
+        "live { " ^ IntSet.fold (fun i s -> s ^ Format.sprintf "%d " i) fields "" ^ "}"
+    | Top -> "top"
+    | Dead -> "dead"
+
+  let print_uses uses =
+    Format.eprintf "Usages:\n";
+    Array.iteri
+      ~f:(fun i ds ->
+        Format.eprintf "v%d: { " i;
+        Var.Map.iter
+          (fun d k ->
+            Format.eprintf
+              "(%a, %s) "
+              Var.print
+              d
+              (match k with
+              | Compute -> "C"
+              | Propagate -> "P"))
+          ds;
+        Format.eprintf "}\n")
+      uses
+
+  let print_liveness live_vars =
+    Format.eprintf "Liveness:\n";
+    Array.iteri ~f:(fun i l -> Format.eprintf "v%d: %s\n" i (live_to_string l)) live_vars
+
+  let print_live_tbl live_table =
+    Format.eprintf "Liveness with dependencies:\n";
+    Var.Tbl.iter
+      (fun v l -> Format.eprintf "%a: %s\n" Var.print v (live_to_string l))
+      live_table
+end
+
+(** Add a sentinal variable declaration to the IR. The fresh variable is assigned to `undefined`. *)
+let add_sentinal p =
+  let sentinal = Var.fresh_n "sentinal" in
+  let undefined = Prim (Extern "%undefined", []) in
+  let instr, loc = Let (sentinal, undefined), Before 0 in
```

Comment:
Should this be `noloc` ?

Reply:
I'm not sure. Right now that definition is inserted as the first line of the program. I think that makes sense in case we need to zero out anything following that.

Reply:
I''ve pushed a commit to use `noloc` (instead of `Before 0`)